### ISSUE TITLE
Resource last_updated does not change when metadata is changed

### ISSUE THREAD

**Issue by ThrawnCA** [ckan/ckan#4951]
The `package_update` action calls:

    `model.Session.query(model.Package).filter_by(id=pkg.id).update(
        {"metadata_modified": datetime.datetime.utcnow()})`

but the `resource_update` action has no equivalent. Updating resource fields will not refresh the 'Last Updated' field.

### CKAN Version if known (or site URL)

2.8.3, 2.9 as at 2019-08-27 on demo.ckan.org

### Please describe the expected behaviour

Updating the metadata of a resource, particularly the URL, should refresh the 'Last Updated' field.

### Please describe the actual behaviour

Only uploading a new file will refresh the 'Last Updated' field.

### What steps can be taken to reproduce the issue? 

Log in to demo.ckan.org

Visit a resource created before today.

Update the resource description.

Check the Last Updated field on the resource page.

**Comment by amercader** [comment]
You are right this is confusing. `last_updated` only refers to when the actual data file was updated (and is only set on uploads). A `metadata_modified` for resources sounds like a good idea if you are up for a PR

**Comment by ThrawnCA** [comment]
Independently of a separate `metadata_modified` field - if `last_updated` shows when a file was uploaded, shouldn't it also be refreshed for URL-based resources when the URL is changed?

**Comment by iamarnavgarg** [comment]
Hi @amercader, I tried to reproduce this issue on my CKAN v2.8 and found that the last modified does change, after performing the reproduction steps- adding a description or changing the resource-url.

Below is the screenshot for the same:
![image](https://user-images.githubusercontent.com/36654563/66826249-af1c9000-ef69-11e9-9f42-2ec74262d3f5.png)

**Comment by boykoc** [comment]
_Notes:_

- _links below are based on CKAN Master (2.9)_
- _demo.ckan.org is running ckan 2.8.2_

If I'm [reading the code right, in CKAN 2.9](https://github.com/ckan/ckan/blob/0261bb87d2d63d540c978b24c9ca582af0b6735a/ckan/lib/dictization/model_save.py#L48) you should be able to use the resource `metadata_modified` field for any metadata updates, including urls.

I've tried on demo.ckan.org:

1. Updating a previously created resource that was a [Link](https://demo.ckan.org/dataset/test-submission-1/resource/d6c55b3f-f454-4338-8f33-aa110fbb6397)
2. Creating then Updating a resource that was an [Upload](https://demo.ckan.org/dataset/test-submission-1/resource/d4a262f9-ca88-4e88-aef8-79da8e6a270e)

In both cases I've found the `last_modified` field on the resource is `null` when viewed by the API and `unknown` when viewed from the web UI.

The screenshot above seems to be from the package (dataset) page instead of the resource page's Additional Information section. On a resource page, the field is displayed as ["Data last updated"](https://github.com/ckan/ckan/blob/0261bb87d2d63d540c978b24c9ca582af0b6735a/ckan/templates/package/resource_read.html#L186).  Looks like Issue #5190 and PR #5236  implemented a [`metadata_modified`](metadata_modified) field for resources but it's also `null` on the demo site (ckan version).

Updating a resource description (Linked or Uploaded) updates the Package (dataset) "Last updated" field.  Actually, just clicking the update button on a resource does this. 

I Followed the reproduction steps on this and did not get the "Data last updated" or "Metadata last updated" fields to update.

For resources, the "last_modified" field is only [updated on Resource Uploads](https://github.com/ckan/ckan/blob/0261bb87d2d63d540c978b24c9ca582af0b6735a/ckan/lib/uploader.py#L225).  

As @amercader mentioned, if the resource is just a link, updating a description or changing the url does not seem to update the `last_updated` field on the resource.  I can't see any logic in the action that would update it either but it may be put somewhere else?

**So, to sum up**. I think this issue is mostly addressed in CKAN 2.9. But I'm unsure why an uploaded resource on ckan 2.8.2 (demo site as of today) is not setting the `last_updated` field for the resource.

**Comment by amercader** [comment]
@boykoc The code that sets last_updated lives in the upload module (ckan/lib/uploader.py). demo.ckan.org uses one of the cloud storage extensions that override the uploader class, and they probably don't implement the last_updated field handling.
As you said on 2.9 you should be able to track the last modified date reliably with the metadata_modified field.

**Comment by boykoc** [comment]
Awesome, thanks for the additional info/explanation on what demo.ckan.org might be doing.
@ThrawnCA, if you're satisfied maybe you could close this issue?

**Comment by ThrawnCA** [comment]
@boykoc That does make sense - although since we do use a cloud storage extension, we should probably look into fixing it there. Thanks.
